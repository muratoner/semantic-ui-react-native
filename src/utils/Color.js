export default class {
	static rgbToYIQ({r, g, b}) {
		return (r * 299 + g * 587 + b * 114) / 1000
	}

	static getBgColor(color) {
		switch (color) {
		case 'primary':
		case 'blue':
			return '#2185d0'
		case 'secondary':
			return '#1b1c1d'
		case 'red':
			return '#db2828'
		case 'orange':
			return '#f2711c'
		case 'yellow':
			return '#fbbd08'
		case 'olive':
			return '##b5cc18'
		case 'green':
			return '#21ba45'
		case 'teal':
			return '#00b5ad'
		case 'violet':
			return '#6435c9'
		case 'purple':
			return '#a333c8'
		case 'pink':
			return '#e03997'
		case 'brown':
			return '#a5673f'
		case 'grey':
			return '#767676'
		case 'black':
			return '#27292a'
		default:
			return color
		}
	}

	static hexToRgb(hex) {
		if (!hex) {
			return undefined
		}

		const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex)

		return result
			? {
				r: parseInt(result[1], 16),
				g: parseInt(result[2], 16),
				b: parseInt(result[3], 16)
			  }
			: undefined
	}

	static contrast(colorHex, threshold = 128) {
		if (!colorHex) {
			return 'rgba(0,0,0,.6)'
		}

		const rgb = this.hexToRgb(colorHex)

		if (rgb === undefined) {
			return 'rgba(0,0,0,.6)'
		}

		return this.rgbToYIQ(rgb) >= threshold ? 'rgba(0,0,0,.6)' : '#fff'
	}

	static stringToColour(str) {
		var hash = 0
		for (var i = 0; i < str?.length; i++) {
			hash = str.charCodeAt(i) + ((hash << 5) - hash)
		}
		var colour = '#'
		for (var i = 0; i < 3; i++) {
			var value = (hash >> (i * 8)) & 0xff
			colour += ('00' + value.toString(16)).substr(-2)
		}
		return colour
	}

	static lightenColor(hex, amt) {
		// From https://css-tricks.com/snippets/javascript/lighten-darken-color/
		var usePound = false
		if (hex[0] === '#') {
			hex = hex.slice(1)
			usePound = true
		}
		var num = parseInt(hex, 16)
		var r = (num >> 16) + amt
		if (r > 255) r = 255
		else if (r < 0) r = 0
		var b = ((num >> 8) & 0x00ff) + amt
		if (b > 255) b = 255
		else if (b < 0) b = 0
		var g = (num & 0x0000ff) + amt
		if (g > 255) g = 255
		else if (g < 0) g = 0
		return (usePound ? '#' : '') + (g | (b << 8) | (r << 16)).toString(16)
	}
}
