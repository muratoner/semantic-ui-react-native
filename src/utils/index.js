import UtilColor from './Color';

export { UtilColor };
