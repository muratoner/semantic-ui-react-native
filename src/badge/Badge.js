import React, {useMemo} from 'react'
import {StyleSheet, Text, View} from 'react-native'
import {UtilColor} from '../utils'

export default (props) => {
	let {
		containerStyle,
		contentStyle,
		style,
		size = 10,
		circular = true,
		color = UtilColor.getBgColor('green'),
		content,
		rounded,
		overlap,
		children,
		...attrs
	} = props
	let computedStyle = useMemo(() => {
		let res = {backgroundColor: UtilColor.getBgColor(color), zIndex: 9}
		if (children) {
			res.position = 'absolute'
			if (overlap != 'circular') {
				res.right = -3
				res.top = -3
			} else {
				res.right = 5
				res.top = 5
			}
		}
		if (typeof content != 'string') {
			res.width = res.height = size
		}
		rounded && (res.borderRadius = 3)
		circular && (res.borderRadius = size * 2)
		return res
	}, [size, color, children, content, circular])
	let getBadge = () => {
		return (
			<View style={[computedStyle, styles.container, style]}>
				{typeof content == 'string' ? (
					<Text
						numberOfLines={1}
						adjustsFontSizeToFit
						style={[
							styles.content,
							{color: UtilColor.contrast(computedStyle.backgroundColor)},
							contentStyle
						]}
					>
						{content}
					</Text>
				) : (
					content
				)}
			</View>
		)
	}
	return (
		<View {...attrs} style={[containerStyle]}>
			{getBadge()}
			{children}
		</View>
	)
}

const styles = StyleSheet.create({
	container: {
		justifyContent: 'center',
		alignItems: 'center'
	},
	content: {
		fontSize: 9,
		textAlign: 'center',
		paddingHorizontal: 4,
		paddingVertical: 2
	}
})
