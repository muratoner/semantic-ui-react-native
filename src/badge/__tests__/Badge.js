import {shallow} from 'enzyme'
import toJson from 'enzyme-to-json'
import React from 'react'
import {StyleSheet} from 'react-native'
import {Avatar} from '../..'
import Icon from '../../icons/Icon'
import {UtilColor} from '../../utils'
import Badge from '../Badge'

describe('Badge Component', () => {
	it('should render without issues', () => {
		const component = shallow(<Badge />)

		expect(component.length).toBe(1)

		expect(component.get(0).props.style).toEqual([undefined])
		expect(
			StyleSheet.flatten(component.get(0).props.children[0].props.style)
		).toHaveProperty('height', 10)
		expect(
			StyleSheet.flatten(component.get(0).props.children[0].props.style)
		).toHaveProperty('height', 10)
		expect(
			StyleSheet.flatten(component.get(0).props.children[0].props.style)
		).toHaveProperty('borderRadius', 20)
		expect(
			StyleSheet.flatten(component.get(0).props.children[0].props.style)
		).toHaveProperty('backgroundColor', UtilColor.getBgColor('green'))

		expect(toJson(component)).toMatchSnapshot()
	})

	it('should render with overlap prop and Avatar child', () => {
		const component = shallow(
			<Badge overlap='circular'>
				<Avatar title='Murat ÖNER' />
			</Badge>
		)

		expect(component.length).toBe(1)
		expect(component.get(0).props.children.length).toBe(2)

		expect(
			StyleSheet.flatten(component.get(0).props.children[0].props.style)
		).toHaveProperty('position', 'absolute')
		expect(
			StyleSheet.flatten(component.get(0).props.children[0].props.style)
		).toHaveProperty('right', 5)
		expect(
			StyleSheet.flatten(component.get(0).props.children[0].props.style)
		).toHaveProperty('top', 5)

		expect(toJson(component)).toMatchSnapshot()
	})

	it('should render with rounded Avatar child', () => {
		const component = shallow(
			<Badge>
				<Avatar title='Murat ÖNER' />
			</Badge>
		)

		expect(component.length).toBe(1)
		expect(component.get(0).props.children.length).toBe(2)

		expect(
			StyleSheet.flatten(component.get(0).props.children[0].props.style)
		).toHaveProperty('position', 'absolute')
		expect(
			StyleSheet.flatten(component.get(0).props.children[0].props.style)
		).toHaveProperty('right', -3)
		expect(
			StyleSheet.flatten(component.get(0).props.children[0].props.style)
		).toHaveProperty('top', -3)

		expect(toJson(component)).toMatchSnapshot()
	})

	it('should render with content', () => {
		const component = shallow(<Badge content={<Icon name='user' />} />)

		expect(component.length).toBe(1)
		expect(component.get(0).props.children[1]).toEqual(undefined)

		expect(toJson(component)).toMatchSnapshot()
	})

	it('should render with rounded', () => {
		const component = shallow(<Badge rounded content='New' />)

		const text = component.find('Text').shallow()

		expect(component.length).toBe(1)
		expect(component.get(0).props.children[1]).toEqual(undefined)
		expect(text.get(0).props.children).toBe('New')

		expect(toJson(component)).toMatchSnapshot()
	})
})
