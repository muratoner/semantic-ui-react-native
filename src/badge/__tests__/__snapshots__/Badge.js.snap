// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge Component should render with content 1`] = `
<View
  style={
    Array [
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#21ba45",
          "borderRadius": 20,
          "height": 10,
          "width": 10,
          "zIndex": 9,
        },
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <Icon
      name="user"
    />
  </View>
</View>
`;

exports[`Badge Component should render with overlap prop and Avatar child 1`] = `
<View
  style={
    Array [
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#21ba45",
          "borderRadius": 20,
          "height": 10,
          "position": "absolute",
          "right": 5,
          "top": 5,
          "width": 10,
          "zIndex": 9,
        },
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  />
  <Component
    title="Murat ÖNER"
  />
</View>
`;

exports[`Badge Component should render with rounded 1`] = `
<View
  style={
    Array [
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#21ba45",
          "borderRadius": 20,
          "zIndex": 9,
        },
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <Text
      adjustsFontSizeToFit={true}
      numberOfLines={1}
      style={
        Array [
          Object {
            "fontSize": 9,
            "paddingHorizontal": 4,
            "paddingVertical": 2,
            "textAlign": "center",
          },
          Object {
            "color": "#fff",
          },
          undefined,
        ]
      }
    >
      New
    </Text>
  </View>
</View>
`;

exports[`Badge Component should render with rounded Avatar child 1`] = `
<View
  style={
    Array [
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#21ba45",
          "borderRadius": 20,
          "height": 10,
          "position": "absolute",
          "right": -3,
          "top": -3,
          "width": 10,
          "zIndex": 9,
        },
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  />
  <Component
    title="Murat ÖNER"
  />
</View>
`;

exports[`Badge Component should render without issues 1`] = `
<View
  style={
    Array [
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#21ba45",
          "borderRadius": 20,
          "height": 10,
          "width": 10,
          "zIndex": 9,
        },
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  />
</View>
`;
